uk_geo_utils/base_updater.py

```diff
@@ -0,0 +1,377 @@
+import abc
+import shutil
+import tempfile
+import urllib.request
+import zipfile
+from pathlib import Path
+
+import psutil
+from django.core.management.base import BaseCommand
+from django.db import ProgrammingError, connection, transaction
+
+
+def unzip(filepath):
+    zip_file = zipfile.ZipFile(filepath, "r")
+    tmpdir = tempfile.mkdtemp()
+    zip_file.extractall(tmpdir)
+    return tmpdir
+
+
+def check_memory(required_memory: int = 2):
+    # Downloading, unzipping and working with the ONSPD
+    # requires a decent chunk of memory to play with.
+    # Running this import on a tiny instance like a
+    # t2.micro will cause an Out Of Memory error
+
+    # By default ensure we've got >2Gb total before we start
+    mem = psutil.virtual_memory()
+    gb = ((mem.total / 1024) / 1024) / 1024
+    return gb >= required_memory
+
+
+# noinspection SqlNoDataSourceInspection
+class BaseUpdater(BaseCommand):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.foreign_key_constraints = None
+        self.indexes = None
+        self.primary_key_constraint = None
```

Comment:
Django models can define arbitrary constraints
https://docs.djangoproject.com/en/4.2/ref/models/constraints/
Just to be clear: If I put some other constraint on my application's local model (extending one of the abstract models in here) what happens to it?
Assuming the answer is: "It gets dropped", how hard would it be to add a pre-flight check for any non-primary/fk constraints somewhere early in the process and `raise Exception("I can't let you do that, Dave")` if we find any?

Reply:
That would be easy to do for the constraints that are in the `pg_constraint` table:

This covers the following constraint types: 
> c = check constraint, f = foreign key constraint, n = not-null constraint (domains only), p = primary key constraint, u = unique constraint, t = constraint trigger, x = exclusion constraint

So it would be easy to check if anything was returned by something like: 

```sql
SELECT 1
FROM pg_constraint
WHERE
    contype NOT IN ('p', 'f')
    AND
    conrelid = '{self.table_name}'::regclass;
```

Reply:
Notes from call: Lets do this, but there is an additional blind spot. Anything that is added via an ALTER TABLE could still be lost.